<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [132]:
import pandas as pd
import psycopg2

In [133]:
n = 10
query = f'''select *
           from public.project_sql
           limit {n}
        '''

In [187]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [136]:
query_3_1 = f'''select
                count(id)
                from VACANCIES
             '''

In [137]:
df = pd.read_sql_query(query_3_1, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [138]:
query_3_2 = f'''select
                count(id)
                from employers
             '''


In [139]:
df = pd.read_sql_query(query_3_2, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [140]:
query_3_3 = f'''select
                count(id)
                from areas
             '''

In [141]:
df = pd.read_sql_query(query_3_3, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [142]:
query_3_4 = f'''select
                count(id)
                from industries
             '''

In [143]:
df = pd.read_sql_query(query_3_4, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

Данные хранят информацию о 49197 вакансий от 23501 работодателей в 1362 регионах мира. 
Количество сфер деятельности работодателей 294.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [144]:
query_4_1 = f'''select   
                     ar.name area,
                     count(v.id) cnt
                from vacancies v JOIN areas ar ON v.area_id=ar.id
                group by ar.name
                order by cnt desc
                LIMIT 10
             '''

In [145]:
df = pd.read_sql_query(query_4_1, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [146]:
query_4_2 = f'''select   
                     count(id) cnt
                from vacancies 
                where salary_from is not null or salary_to is not null
             '''

In [147]:
df = pd.read_sql_query(query_4_2, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [148]:
query_4_3 = f'''select   
                     AVG(salary_from),
                     AVG(salary_to)
                from vacancies 
             '''

In [149]:
df = pd.read_sql_query(query_4_3, connection)
round(df)

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg       avg
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [150]:
query_4_4 = f'''select   
                     v.schedule,
                     v.employment,
                     count(v.id)
                from vacancies v 
                group by v.schedule, v.employment
                order by 3 desc
                
             '''

In [151]:
df = pd.read_sql_query(query_4_4, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [152]:
query_4_5 = f'''select   
                     v.experience,
                     count(v.id)
                from vacancies v 
                group by v.experience
                order by 2
                
             '''

In [153]:
df = pd.read_sql_query(query_4_5, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Наибольшее количество вакансий находятся в столицах стран СНГ и городах-миллиониках России, связанно это несомнено с высоким и динамичным экономическим развитием данных регионов и большей потребности в новых сотрудниках.       
Информация об уровне заработной платы предоставлена в 24073 из 49197, что может быть связанно с желанием работодателя обсудить уровень заработной платы желаемый соискателем после собеседования.    
Среднее значение вилки заработной платы 71065	и 110537, данный признак считаю не информативным из-за большого количества регионов в анализе. Разница между средними зарабатными платами в разных регионах, как правило, велика и учитывает среднее по незаполненым данным, что видно из следующего дополнительного анализа. 



In [154]:
query_4_6 = f'''SELECT
                     ar.name, 
                     AVG(v.salary_from),    
                     AVG(v.salary_to)
                FROM vacancies v 
                     JOIN areas ar ON v.area_id=ar.id 
   WHERE salary_from IS NOT NULL    /*убираем из анализа вакансии без указания верхнего и нижнего порога зп*/
      AND salary_to IS NOT NULL
GROUP BY ar.name         /*создаем выборку средней зп по регионам*/
ORDER BY 3 DESC
LIMIT 10 '''
df = pd.read_sql_query(query_4_6, connection)
round(df)

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name       avg       avg
0              Германия  348270.0  557232.0
1               Кутаиси  160734.0  350691.0
2            Черногория  282501.0  331208.0
3               Кинешма  150000.0  300000.0
4              Болгария  200000.0  300000.0
5           Таджикистан  233794.0  292242.0
6               Кишинёв  233794.0  292242.0
7                Сербия  219182.0  277630.0
8                  Кипр  207225.0  272206.0
9  Дзержинск (Беларусь)  171838.0  268860.0

Из анализа ТОП 10 регионов по уровню вилки средней заработной платы мы видим, что предложения работодадетелей значительно отличаются от региона к региону.     
Анализ рабочего дня и типа занятости показывает, что наиболее популярен запрос работодателей на полный день и полную занятость. На втором и третьем месте удаленная работа и гибкий график, но так же предполагающая полную занятость.     
 Чаще всего работодатель заинтересован в соискателях имеющих опыт работы от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [155]:
query_5_1 = f'''select   
                     em.name employer,
                     count(v.id)  count_vacancies
                from employers em 
                     JOIN vacancies v ON v.employer_id = em.id 
                group by em.name
                order by 2 desc
                
             '''

In [156]:
df = pd.read_sql_query(query_5_1, connection)
df.head()

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer  count_vacancies
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [157]:
query_5_2 = f'''SELECT
                      a.name, 
                     count (distinct e.id) count_employers, 
                     count(v.id) count_vacancies 
                FROM areas a
                left join employers e ON a.id = e.area
                left join vacancies v ON a.id = v.area_id
                GROUP BY a.id
                HAVING count(v.id)=0 
                ORDER BY 2 desc
                LIMIT 1 
             '''

In [158]:
df = pd.read_sql_query(query_5_2, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count_employers  count_vacancies
0  Россия              410                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [159]:
query_5_3 = f'''SELECT
                    distinct e.name,  
                    count(distinct v.area_id) count_area
                FROM employers e
                left join vacancies v ON e.id = v.employer_id
                GROUP BY e.name
                ORDER BY 2 desc
             '''

In [160]:
df = pd.read_sql_query(query_5_3, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count_area
0                       Яндекс         181
1                   Ростелеком         152
2                   Спецремонт         116
3       Поляков Денис Иванович          88
4                    ООО ЕФИН           71
...                        ...         ...
23170             Ярпож Казань           0
23171                   ЯРСНИП           0
23172   Ясли-сад № 28 г.Минска           0
23173            Яшин&Партнёры           0
23174  Яшнова Алена Викторовна           0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [161]:
query_5_4 = f'''SELECT
                    count(e.id)
                FROM employers e
                LEFT JOIN employers_industries ei ON e.id=ei.employer_id
                WHERE industry_id IS NULL 
             '''

In [162]:
df = pd.read_sql_query(query_5_4, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [163]:
query_5_5 = f'''SELECT e.name, 
                     count(ei.industry_id) 
                FROM public.employers e 
                LEFT JOIN public.employers_industries ei ON e.id=ei.employer_id 
                GROUP BY e.name
                HAVING count(ei.industry_id)=4
                ORDER BY e.name
                OFFSET 2 LIMIT 1
             '''

In [164]:
df = pd.read_sql_query(query_5_5, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [165]:
query_5_6 = f'''SELECT i.name, count(e.name)
                FROM public.employers e 
                LEFT JOIN public.employers_industries ei ON e.id=ei.employer_id 
                LEFT JOIN public.industries i ON ei.industry_id=i.id 
                WHERE i.name IS NOT NULL 
                GROUP BY i.name 
                HAVING i.name='Разработка программного обеспечения'
             '''

In [166]:
df = pd.read_sql_query(query_5_6, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0  Разработка программного обеспечения   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [167]:
query_5_7 = f'''  SELECT
                       a.name,
                       COUNT(v.id) cnt
                  FROM public.areas a                                                                                                        
                  JOIN public.vacancies v ON a.id=v.area_id
                  JOIN public.employers e ON e.id=v.employer_id
                  WHERE e.name = 'Яндекс' AND a.name IN ('Москва', 'Санкт-Петербург','Новосибирск', 
                  'Екатеринбург','Казань', 'Нижний Новгород', 'Челябинск', 
                  'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск','Краснодар', 
                  'Воронеж', 'Пермь', 'Волгоград')                                                                                          
                  GROUP BY a.name
                  UNION ALL
                  SELECT 
                  'Total',
                  COUNT(*)
                  FROM public.areas a
                  JOIN public.vacancies v ON a.id=v.area_id
                  JOIN public.employers e ON e.id=v.employer_id
                  WHERE e.name ='Яндекс' AND a.name IN ('Москва', 'Санкт-Петербург','Новосибирск', 
                  'Екатеринбург','Казань', 'Нижний Новгород', 'Челябинск', 
                  'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск','Краснодар', 
                  'Воронеж', 'Пермь', 'Волгоград')
                  ORDER BY 2'''

In [168]:
df = pd.read_sql_query(query_5_7, connection)
df.shape[0]

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


17

In [169]:
df = pd.read_sql_query(query_5_7, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

ТОП 5 работодадетелй по количеству вакансий: Яндекс	, Ростелеком, Тинькофф, СБЕР, Газпром нефть.	
ИЗ регионов, в которых нет вакансий, в регион "Россия"  наибольшее количество работодателей, связанно это с некорректными данными, т.к. каждый регион РФ заведен отдельно. Это видно из дополнительного запроса по количеству работодателей и вакансий в регионах

In [170]:
query_5_8 = f'''SELECT
                      a.name, 
                     count (distinct e.id) count_employers, 
                     count(v.id) count_vacancies 
                FROM areas a
                left join employers e ON a.id = e.area
                left join vacancies v ON a.id = v.area_id
                GROUP BY a.id
                ORDER BY 2 desc
                LIMIT 10
             '''
df = pd.read_sql_query(query_5_8, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count_employers  count_vacancies
0           Москва             5864         31272712
1  Санкт-Петербург             2217          6320667
2            Минск             1115          2354880
3           Алматы              721          1364132
4     Екатеринбург              609          1034082
5      Новосибирск              573          1149438
6           Казань              480           679200
7  Нижний Новгород              426           711420
8           Россия              410                0
9        Краснодар              409           532109

Лидирующие позиции по количеству вакансий и работодателей занимаю вполне российские города: Москва, Санкт-Петербург, Екатеринбург и т.д. Необходима корректировка занесения данных, т.к. это может повлечь ошибки в анализе.    
Работодатели наиболее широко представленные по регионам: Яндекс, Ростелеком, Спецремонт, Поляков Денис Иванович, ООО ЕФИН, что позволяет сделать вывод о возможно удаленной работе в этих компаниях, либо широкой географии работы компании.     
У 8419 из 23501 работодателей не предоставлена сфера деятельности, это 35%. Необходимо дополнительный анализ количества откликов на вакансии данных работодателей и алгоритма их рекомендации соискателям.    
Для выявления работодателей с наибольшим количеством видов деятельности создан дополнительный запрос ниже:

In [171]:
query_5_9 = f'''SELECT e.name, 
                     count(ei.industry_id) 
                FROM public.employers e 
                LEFT JOIN public.employers_industries ei ON e.id=ei.employer_id 
                GROUP BY e.name
                ORDER BY 2 DESC
                LIMIT 10 
             '''
df = pd.read_sql_query(query_5_9, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0            Модуль     16
1   Простые решения     11
2           Энергия     10
3             Прайд     10
4               СБК      9
5             Альфа      9
6           Эксперт      9
7  Бизнес-Ассистанс      8
8            Уфанет      8
9        Автомастер      8

В качестве вида деятельности "разработка ПО" указана у 3553 работодателей. У крупнейшего работодателя Яндекс, только 485 из 1933 вакансий размещены в городах-миллионниках.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [172]:
query_6_1 = f'''SELECT count(name)
                FROM vacancies
                WHERE name iLIKE '%DATA%'
                OR name iLIKE '%ДАНН%' 
                '''

In [173]:
df = pd.read_sql_query(query_6_1, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [174]:
query_6_2 = f'''SELECT count(id)
                FROM vacancies v
                WHERE ((name ILIKE '%data scientist%') 
                OR (name ILIKE '%data science%') 
                OR (name ILIKE '%исследователь данных%')
                OR ((name ILIKE '%ML%')
                AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL))
                AND (v.name ILIKE '%junior%' 
                OR v.experience ILIKE '%Нет опыта%' 
                OR v.employment ILIKE '%Стажировка%')
                '''

In [175]:
df = pd.read_sql_query(query_6_2, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [176]:
query_6_3 = f'''SELECT count(id)
                FROM vacancies v
                WHERE ((name ILIKE '%data scientist%') 
                OR (name ILIKE '%data science%') 
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%')
                AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL))
                AND (key_skills ILIKE '%SQL%' OR key_skills ILIKE 'postgres')
                '''

In [177]:
df = pd.read_sql_query(query_6_3, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [178]:
query_6_4 = f'''SELECT count(id)
                FROM vacancies v
                WHERE ((name ILIKE '%data scientist%') 
                OR (name ILIKE '%data science%') 
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%')
                AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL))
                AND key_skills ILIKE '%Python%' 
                '''

In [179]:
df = pd.read_sql_query(query_6_4, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [180]:
query_6_5 = f'''SELECT ROUND(AVG(length(key_skills) - length(REPLACE(key_skills, CHR(9), '' ))+1)::numeric, 2) 
                FROM public.vacancies v 
                WHERE ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') 
                OR (name ILIKE '%исследователь данных%') 
                OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%')) 
                OR (name ILIKE '%machine learning%') 
                OR (name ILIKE '%машинн%обучен%') 
                AND (key_skills IS NOT NULL))'''

In [181]:
df = pd.read_sql_query(query_6_5, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [182]:
query_6_6 = f'''SELECT
                     experience,
                    ROUND(AVG(COALESCE((salary_from + salary_to)/2, salary_from, salary_to)),0)
                    FROM vacancies
                    WHERE ((name ILIKE '%data scientist%') 
                    OR (name ILIKE '%data science%')  
                    OR (name ILIKE '%исследователь данных%')    
                    OR (name LIKE '%ML%') 
                    AND (name NOT ILIKE '%HTML%')) 
                    OR (name ILIKE '%machine learning%') 
                    OR (name ILIKE '%машинн%обучен%') 
                    AND ((salary_from IS NOT NULL) OR (salary_to IS NOT NULL))
                    group by experience'''

In [183]:
df = pd.read_sql_query(query_6_6, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

Из предметного анализа данных следует, что 1771 вакансия так или иначе имеет отношение к работе с данными. При этом 51 из них для специалистов  DS начального уровня.    
Для поиска количества вакансий специалистов DS любого уровня необходим дополнительный запрос:   

In [184]:
query_6_7 = f'''SELECT count(id)
                FROM vacancies v
                WHERE ((name ILIKE '%data scientist%') 
                OR (name ILIKE '%data science%') 
                OR (name ILIKE '%исследователь данных%')
                OR ((name ILIKE '%ML%')
                AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%'))
             '''
df = pd.read_sql_query(query_6_7, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    536

Всего размещено 536 вакансий: из них в 201 в качестве ключевого навыка указан SQL или postgres, Python необходим в 351, т.е более чем в 65% случаев, что позволяет сделать вывод о том, что это один из самых важных и ключевых навыков специаллиста DS. В среднем работодатели указывают 6-7 ключевых навыков для специалиста DS. Специалист без опыта работы в DS согласно размещенным вакансиям может расчитывать на зп около 75000, тогда специалист с опытом работы от 3 до 6 лет вполне может получать более 250 000 руб.

# Общий вывод по проекту

Был проведен анализ 49197 вакансий от 23501 работодателей в 1362 регионах.
Информация об уровне заработной платы предоставлена в 24073 из 49197, что, как я уже отмечала, может быть связанно с желанием работодателя обсудить уровень заработной платы на собеседовании. Необходимо провести дополнительный анализ уровня откликов соискателей на вакансии без указания уровня дохода, анализ вакансий и работодателей, которые чаще всего не указывают уровень зп. Так же данный факт необходимо учитывать при построении алгоритма рекомендации соискателям таких вакансий. Рассмотреть возможность исключения пустых данных из анализа, либо возможность обязательного заполнения данной информации.   
Из-за большого количества регионов в анализе уровень средней вилки по зп не информативен, проведен дополнительный анализ регионов из которого видно, что предложения работодадетелей значительно отличаются от региона к региону. 
Для построения алгоритма рекомендации работодаделям по зп необходим дополнительный анализ предлагаемой зп по регионам, видам деятельности и предметный анализ аналогичных вакансий.         
Выявлены самые популярные запросы работодателей по типу занятости, графику работы, опыту.	
В анализе вакансий по регионам обнаружена проблема занесения информации в БД, т.к не всегда указывается город или страна. Необходима корректировка занесения данных, т.к. это может повлечь ошибки в анализе.    
В ходе анализа выявлены работодатели наиболее широко представленные по разным регионам.    
Необходим дополнительный анализ количества откликов на вакансии работодателей с непредставленным видом деятельности, т.к. таких 35% это может значительно сказаться на дальнейшем анализе данных. 
Вид деятельности разработка ПО указана у 15% работодателей. 
3% от всех вакансий имеет отношение к работе с данными, из них 30% для специалистов DS разных уровней. В качестве ключевых навыков популярны Python, SQL. Необходимо найти самые популярные навыки для специалиста DS:

In [185]:
query_6_8 = f'''SELECT key_skills,
                       count(key_skills)
                FROM vacancies v
                WHERE ((name ILIKE '%data scientist%') 
                OR (name ILIKE '%data science%') 
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%')
                AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL))
                AND key_skills ILIKE '%Python%' 
                group By key_skills
                ORDER BY 2 DESC
                LIMIT 10''' 
df = pd.read_sql_query(query_6_8, connection)
df

c:\Users\oksaz\anaconda3\lib\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


key_skills  count
0  Python\tMachine Learning\tData Analysis\tNumpy...     14
1  Python\tmachine learning\tweb security\tsklear...      9
2  Python\tSQL\tData Mining\tАналитические исслед...      6
3        Python\tLearning to Rank\tData Science\tNLP      6
4  Python\tDocker\tMachine Learning\tData Science...      5
5   Python\tSQL\tNLP\tScikit-learn\tMachine Learning      4
6  Python\tC++\tАнглийский язык\tMachine Learning...      3
7  Python\tdata science\tРазработка поисковых тех...      3
8  Machine Learning\tDeep Learning\tData Science\...      3
9  Python\tJava\tC++\tАлгоритмы и структуры данны...      3

Это при дальнейшем анализе, позволит выявить какие навыки необходимо приобрести специалисту DS и указать в резюме.    
Выявлен уровень зп в зависимости от опыта работы по специальности.

In [186]:
connection.close()